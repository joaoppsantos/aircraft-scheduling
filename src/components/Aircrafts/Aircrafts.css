.aircrafts {
    height: calc(100vh - 15rem);
    width: calc(25vh);
    padding-top: 10px;
    background-color: lavender;
    margin-right: 20px;
}

@media screen and (max-width: 670px) {
	.aircrafts {
        height: calc(100vh - 15rem);
        width: 90%;
        padding-top: 10px;
        background-color: lavender;
        margin: 0 auto;
	}
}