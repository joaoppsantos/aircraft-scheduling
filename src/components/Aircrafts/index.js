export { default as Aircrafts } from './Aircrafts';
