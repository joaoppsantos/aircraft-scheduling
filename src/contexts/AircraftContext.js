import { createContext, useEffect, useState } from 'react';
import request from '../request';

export const AircraftContext = createContext({});

export function AircraftProvider({ children }) {
	const [aircrafts, setAircrafts] = useState([]);
	const [isAircraftLoading, setIsAircraftLoading] = useState(true);
	const [currentAircraft, setCurrentAircraft] = useState(undefined);
	const [currentPercentage, setCurrentPercentage] = useState(0);

	const setPercentage = percentage => {
		setCurrentPercentage(percentage);
	};
	const selectAircraft = aircraft => {
		setCurrentAircraft(aircraft);
	};

	useEffect(() => {
		request.get(`/aircrafts?limit=5&offset=0`).then((response) => {
			setAircrafts(response.data.data);
			setIsAircraftLoading(false);
		});
	}, []);

	return (
		<AircraftContext.Provider value={{
			aircrafts,
			isAircraftLoading,
			currentAircraft,
			selectAircraft,
			currentPercentage,
			setPercentage
		}}>
			{children}
		</AircraftContext.Provider>
	)
}
