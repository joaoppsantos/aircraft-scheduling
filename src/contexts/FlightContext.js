import request from '../request';
import { createContext, useEffect, useState } from 'react';

export function FlightProvider({ children }) {
	const [flights, setFlights] = useState([]);
	const [activeFlights, setActiveFlights] = useState([]);
	const [offset, setOffset] = useState(-1);
	const [hasFlightsRemaining, setHasFlightsRemaining] = useState(true);
	const paginationLimit = 25;

	const toggleActiveFlight = (flight) => {
		let isFlightOnList = false;

		activeFlights.forEach((activeInfo) => {
			if (activeInfo.id === flight.id) {
				isFlightOnList = true;
			}
		});

		isFlightOnList ? 
			setActiveFlights(activeFlights.filter(activeInfo => activeInfo.id !== flight.id))
			: setActiveFlights([...activeFlights, flight]);
	}

	useEffect(() => {
		if (!hasFlightsRemaining) {
			return;
		}

		async function fetchData() {
			try {
				const newOffset = offset + 1;
				const response = await request.get(`/flights?limit=${paginationLimit}&offset=${newOffset}`);

				setOffset(newOffset);
				setFlights([...flights, ...response?.data?.data]);

				if (paginationLimit * (newOffset + 1) > response?.data?.pagination?.total) {
					setHasFlightsRemaining(false);
				}
			} catch (error) {
				console.log('Failed to load !');
			};
		}

		fetchData();
	// eslint-disable-next-line react-hooks/exhaustive-deps
	}, [flights, hasFlightsRemaining]);

	return (
		<FlightContext.Provider value={{
			flights,
			toggleActiveFlight,
			activeFlights
		}}>
			{ children }
		</FlightContext.Provider>
	)
}

export const FlightContext = createContext({});