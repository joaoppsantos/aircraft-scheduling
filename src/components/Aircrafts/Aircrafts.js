import './Aircrafts.css';
import { useContext } from 'react';
import { AircraftCard } from '../AircraftCard';
import { AircraftContext } from '../../contexts/AircraftContext';

const Aircrafts = () => {
	const {
		aircrafts
	} = useContext(AircraftContext);

	return (
		<div className="aircrafts">
			Aircrafts
			{aircrafts.map((aircraft, index) => {
                return <AircraftCard key={index} {...aircraft} />
            })}
		</div>
	);
}

export default Aircrafts;
