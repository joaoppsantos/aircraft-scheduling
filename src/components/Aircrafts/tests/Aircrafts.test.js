import * as React from 'react';
import { render } from '@testing-library/react';
import Aircrafts from '../Aircrafts';
import { AircraftContext } from '../../../contexts/AircraftContext';
import { FlightContext } from '../../../contexts/FlightContext';
import { aircraftMock, flightMock } from '../../../tests/mocks/mocks';

describe('<Aircrafts />', () => {
    it('should render correctly', () => {
        const mockFlightContext = {
            activeFlights: [
                flightMock,
            ],
        };
        const mockAircraftContext = {
            aircrafts: [
                aircraftMock
            ],
            selectAircraft: jest.fn(),
            currentPercentage: "50",
            setPercentage: jest.fn()
        }

        const { container } = render(
            <AircraftContext.Provider value={mockAircraftContext}>
                <FlightContext.Provider value={mockFlightContext}>
                    <Aircrafts />
                </FlightContext.Provider>
            </AircraftContext.Provider>
        );

        expect(container).toMatchSnapshot();
    });
});
