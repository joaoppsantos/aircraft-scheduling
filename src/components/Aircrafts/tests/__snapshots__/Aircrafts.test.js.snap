// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Aircrafts /> should render correctly 1`] = `
<div>
  <div
    class="aircrafts"
  >
    Aircrafts
    <div
      class="aircraft-card"
    >
      <span>
        GABCD
      </span>
      <span
        class="percentage"
      >
        50
        %
      </span>
    </div>
  </div>
</div>
`;
